server/src/main/java/io/druid/segment/realtime/FireHydrant.java

```diff
@@ -59,14 +59,30 @@ public FireHydrant(
     this.count = count;
   }
 
+  public int indexSize() {
+    synchronized (swapLock) {
+      return index == null ? 0 : index.size();
+    }
+  }
+
+  public boolean canAppendRow() {
+    synchronized (swapLock) {
+      return index != null && index.canAppendRow();
+    }
+  }
+
   public IncrementalIndex getIndex()
   {
-    return index;
+    synchronized (swapLock) {
+      return index;
```

Comment:
If index can be null, it's not safe to do something like `return index != null && index.canAppendRow()`. I thought the overhead is minimal. But `volatile` needed to be removed then.